File: backend/app/api/src/main/kotlin/com/wallet/clover/api/controller/CommunityController.kt
package com.wallet.clover.api.controller

import com.wallet.clover.api.common.CommonResponse
import com.wallet.clover.api.common.PageResponse
import com.wallet.clover.api.dto.Comment
import com.wallet.clover.api.dto.CreateComment
import com.wallet.clover.api.dto.CreatePost
import com.wallet.clover.api.dto.Post
import com.wallet.clover.api.dto.UpdateComment
import com.wallet.clover.api.dto.UpdatePost
import com.wallet.clover.api.service.CommunityService
import jakarta.validation.Valid
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/community")
class CommunityController(
    private val communityService: CommunityService,
) {

    @GetMapping("/posts")
    suspend fun getAllPosts(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @AuthenticationPrincipal jwt: Jwt?
    ): CommonResponse<PageResponse<Post.Response>> {
        val userId = jwt?.subject?.toLongOrNull()
        return CommonResponse.success(communityService.getAllPosts(page, size, userId))
    }

    @GetMapping("/posts/{postId}")
    suspend fun getPost(
        @PathVariable postId: Long,
        @AuthenticationPrincipal jwt: Jwt?
    ): CommonResponse<Post.Response> {
        val userId = jwt?.subject?.toLongOrNull()
        return CommonResponse.success(communityService.getPostById(postId, userId))
    }

    @PostMapping("/posts")
    suspend fun createPost(
        @Valid @RequestBody request: CreatePost.Request,
        @AuthenticationPrincipal jwt: Jwt
    ): CommonResponse<Post.Response> {
        val userId = jwt.subject.toLong()
        return CommonResponse.success(communityService.createPost(userId, request))
    }

    @PutMapping("/posts/{postId}")
    suspend fun updatePost(
        @PathVariable postId: Long,
        @Valid @RequestBody request: UpdatePost.Request,
        @AuthenticationPrincipal jwt: Jwt
    ): CommonResponse<Post.Response> {
        val userId = jwt.subject.toLong()
        return CommonResponse.success(communityService.updatePost(postId, userId, request))
    }

    @PostMapping("/posts/{postId}/like")
    suspend fun likePost(
        @PathVariable postId: Long,
        @AuthenticationPrincipal jwt: Jwt
    ): CommonResponse<Post.Response> {
        val userId = jwt.subject.toLong()
        return CommonResponse.success(communityService.likePost(postId, userId))
    }

    @GetMapping("/posts/{postId}/comments")
    suspend fun getComments(
        @PathVariable postId: Long,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int
    ): CommonResponse<PageResponse<Comment.Response>> {
        return CommonResponse.success(communityService.getCommentsByPostId(postId, page, size))
    }

    @PostMapping("/comments")
    suspend fun createComment(
        @Valid @RequestBody request: CreateComment.Request,
        @AuthenticationPrincipal jwt: Jwt
    ): CommonResponse<Comment.Response> {
        val userId = jwt.subject.toLong()
        return CommonResponse.success(communityService.createComment(userId, request))
    }

    @PutMapping("/comments/{commentId}")
    suspend fun updateComment(
        @PathVariable commentId: Long,
        @Valid @RequestBody request: UpdateComment.Request,
        @AuthenticationPrincipal jwt: Jwt
    ): CommonResponse<Comment.Response> {
        val userId = jwt.subject.toLong()
        return CommonResponse.success(communityService.updateComment(commentId, userId, request))
    }
}

File: backend-node/src/community/community.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  UseGuards,
  Request,
  Req,
  ParseIntPipe,
} from '@nestjs/common';
import { CommunityService } from './community.service';
import { CreatePostDto } from './dto/create-post.dto';
import { UpdatePostDto } from './dto/update-post.dto';
import { CreateCommentDto } from './dto/create-comment.dto';
import { UpdateCommentDto } from './dto/update-comment.dto';
import { AuthGuard } from '@nestjs/passport';

/**
 * 커뮤니티 게시글 및 댓글 관리를 담당하는 컨트롤러입니다.
 * Kotlin CommunityController 로직을 이식함.
 */
@Controller('community')
export class CommunityController {
  constructor(private readonly communityService: CommunityService) {}

  /**
   * 모든 게시글을 페이지네이션하여 조회합니다.
   * @param page 페이지 번호
   * @param size 페이지당 게시글 수
   */
  @Get('posts')
  async getAllPosts(
    @Query('page') page = 0,
    @Query('size') size = 10,
    @Request() req: any,
  ) {
    // JWT 토큰이 있으면 사용자 ID 추출 (비로그인 허용)
    const userId = req.user?.id;
    return this.communityService.getAllPosts(+page, +size, userId);
  }

  /**
   * 특정 ID의 게시글 상세 정보를 조회합니다.
   */
  @Get('posts/:id')
  async getPost(@Param('id') id: string, @Request() req: any) {
    const userId = req.user?.id;
    return this.communityService.getPostById(BigInt(id), userId);
  }

  /**
   * 새로운 게시글을 작성합니다.
   */
  @UseGuards(AuthGuard('jwt'))
  @Post('posts')
  async createPost(@Request() req: any, @Body() dto: CreatePostDto) {
    return this.communityService.createPost(req.user.id, dto);
  }

  /**
   * 기존 게시글을 수정합니다.
   */
  @UseGuards(AuthGuard('jwt'))
  @Put('posts/:id')
  async updatePost(
    @Request() req: any,
    @Param('id') id: string,
    @Body() dto: UpdatePostDto,
  ) {
    return this.communityService.updatePost(BigInt(id), req.user.id, dto);
  }

  /**
   * 게시글에 '좋아요'를 표시하거나 취소합니다.
   */
  @UseGuards(AuthGuard('jwt'))
  @Post('posts/:id/like')
  async likePost(@Request() req: any, @Param('id') id: string) {
    return this.communityService.likePost(BigInt(id), req.user.id);
  }

  /**
   * 특정 게시글의 댓글 목록을 조회합니다.
   */
  @Get('posts/:id/comments')
  async getComments(
    @Param('id') id: string,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    return this.communityService.getCommentsByPostId(BigInt(id), +page, +size);
  }

  /**
   * 게시글에 새로운 댓글을 작성합니다.
   */
  @UseGuards(AuthGuard('jwt'))
  @Post('comments')
  async createComment(@Request() req: any, @Body() dto: CreateCommentDto) {
    return this.communityService.createComment(req.user.id, dto);
  }

  /**
   * 기존 댓글을 수정합니다.
   */
  @UseGuards(AuthGuard('jwt'))
  @Put('comments/:id')
  async updateComment(
    @Request() req: any,
    @Param('id') id: string,
    @Body() dto: UpdateCommentDto,
  ) {
    return this.communityService.updateComment(BigInt(id), req.user.id, dto);
  }

  /**
   * 게시글을 삭제합니다. (작성자만 가능)
   */
  @Delete('posts/:id')
  @UseGuards(AuthGuard('jwt'))
  async deletePost(@Param('id', ParseIntPipe) id: number, @Req() req: any) {
    await this.communityService.deletePost(BigInt(id), req.user.id);
    return { message: 'Post deleted' };
  }

  /**
   * 특정 사용자가 작성한 게시글 목록을 조회합니다.
   */
  @Get('users/:userId/posts')
  async getUserPosts(
    @Param('userId', ParseIntPipe) userId: number,
    @Query('page', new ParseIntPipe({ optional: true })) page?: number,
    @Query('size', new ParseIntPipe({ optional: true })) size?: number,
  ) {
    return this.communityService.getUserPosts(BigInt(userId), page ?? 0, size ?? 10);
  }
}
